File: 22-generate-parentheses/22-generate-parentheses.cpp
class Solution {
public:
    vector<string>res;
    vector<string> generateParenthesis(int n) {
        fun("", n, 0);
        return res;
    }
    
    void fun(string s, int u, int left)
    {
        if(u==0 and left ==0 )
        {
            res.push_back(s);
            return;
        }
        if(u>0)
        {
            fun(s+'(', u-1, left+1);
        }
        if(left>0)
        {
            fun(s+')', u, left-1);
        }
    }
};

File: 22-generate-parentheses/22-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        helper(n,0,0,0,"",ans);
        return ans;
    }
    
    public void helper(int n,int i, int open,int close,String str,List<String> ans) {
        if (i==2*n) {
            ans.add(str);
            return;
        }
        
        if (open<n) {
            helper(n,i+1,open+1,close,str+"(",ans);
        }
        
        if (close<open) {
            helper(n,i+1,open,close+1,str+")",ans);
        }
    }
}